File: Sources/NockerlDesign/Components/NockerlRecordingHUD.swift
// NockerlRecordingHUD: the floating "Nockerl is listening" pill on Swift, aligned
// VALUE-EXACT to the web canon (site RecordingHudDemo.tsx) per the design lead's
// ruling: "look just like the web version, but in Swift." Twin of the Compose
// NockerlRecordingHud.
//
// ANATOMY (left → right, showBrand default true):
//   [ NockerlLogo (16, theme ink) ] → [ gray vertical divider ] → [ CONTENT that
//   MORPHS per phase ] → [ optional trailing ghost Cancel ]
//
// THE COLOR SPLIT (the core fix that kills the over-warming, law §10): the record DOT
// is the ONLY warm element (statusError); the TIMER is THEME INK (onCanvas, never
// red); the METER bars are accentPrimary (brand cyan). The pill carries the §2
// floating accent edge (accentPrimary; statusWarning in the error phase only).
//
// PHASES ( folded in, the full dictation lifecycle, not recording-only):
//   recording:    pulsing red dot → mono timer (ink) → live cyan equalizer
//   paused:       static dimmed dot → muted timer → frozen muted equalizer → "Paused"
//   transcribing: indeterminate spinner → "Transcribing…"
//   error:        warn triangle → error text (2-line); pill border → statusWarning
//
// Golden determinism: the dot pulse + the meter are the infinite bits.
// Pass `animate: false` to freeze the dot at peak; the equalizer draws the supplied
// levels directly (a fixed list is deterministic). Per-bar ease is opt-in
// (`animateBars`, default off) at the web's ~80ms.

import SwiftUI

/// The §2 floating-over-content signature border (design-laws §2): any fixed
/// element floating over scrollable content carries this THICK cyan border on
/// its floating PARENT container. Twin of the Compose `NockerlFloatingBorderWidth`.
///
/// v1.17.0 made this an ALIAS onto the generated `NockerlBorder.widthFloating` token rather than
/// a hand-authored 1.5 (the drift the `border.width` token set exists to prevent). Kept as a
/// named public symbol because consumers already call it, and because the NAME is the point:
/// it says "this element floats", which `widthFloating` alone at a call site does not.
public enum NockerlFloatingBorder {
    /// The ratified "thick" weight, matched to the native chat-input pill.
    public static let width: CGFloat = NockerlBorder.widthFloating
}

/// The HUD's dictation phase, mirroring the web `HudPhase`. Drives
/// which content morphs in after the fixed logo + divider, plus the pill's border
/// tone (`error` → statusWarning).
public enum NockerlRecordingHudPhase: Equatable {
    /// Live capture: pulsing red dot → mono timer (ink) → live cyan meter.
    case recording
    /// Held: static dimmed dot → muted timer → frozen muted meter → "Paused".
    case paused
    /// Post-capture: indeterminate spinner → "Transcribing…" (no dot/timer/meter).
    case transcribing
    /// Failed: warn triangle → error text; the pill border rides statusWarning.
    case error
    /// Success: a checkmark → "Pasted"/"Copied" (drops the dot/timer/meter, like
    /// transcribing). The §2 accent edge stays. `pasted` = pasted-vs-copied.
    case result(pasted: Bool)
}

/// The HUD's opt-in entrance/exit ( bonus, additive, and the default `.none`
/// preserves the current in-place appearance). Motion is EYEBALL-only.
public enum NockerlRecordingHudEntrance: Equatable {
    /// No transition: appears/disappears in place (the default, unchanged).
    case none
    /// Pops UP from the bottom screen edge on insert, shoots back DOWN on removal
    /// (slide + fade). The HOST toggles the HUD's presence inside `withAnimation`;
    /// this equips the pill with the bottom-edge transition. Holds still while live.
    case fromBottom
}

/// The floating recording HUD pill (see the file header), value-exact to the web
/// canon. `phase` selects the morphing content; the logo + divider are fixed.
public struct NockerlRecordingHUD: View {
    private let phase: NockerlRecordingHudPhase
    private let elapsedLabel: String
    private let levels: [CGFloat]
    private let errorMessage: String
    private let transcribingLabel: String
    private let pausedLabel: String
    private let resultPastedLabel: String
    private let resultCopiedLabel: String
    private let showBrand: Bool
    private let animate: Bool
    private let animateBars: Bool
    private let showsCancel: Bool
    private let cancelLabel: String
    private let accessibilityLabelOverride: String?
    private let entrance: NockerlRecordingHudEntrance
    private let styleSelector: NockerlHudStyleSelector?
    private let onCancel: () -> Void

    @State private var pulsed = false
    @Environment(\.colorScheme) private var colorScheme
    @Environment(\.accessibilityReduceMotion) private var reduceMotion

    /// Create a recording HUD.
    /// - Parameters:
    ///   - phase: the dictation phase (default `.recording`).
    ///   - elapsedLabel: the pre-formatted elapsed time (`m:ss`) for recording / paused.
    ///   - levels: per-bar levels (0…1) for the equalizer in recording / paused.
    ///   - errorMessage: the failure text (the `.error` phase).
    ///   - transcribingLabel: the transcribing status word.
    ///   - pausedLabel: the paused status word.
    ///   - resultPastedLabel: the success word when `.result(pasted: true)`.
    ///   - resultCopiedLabel: the success word when `.result(pasted: false)`.
    ///   - showBrand: `false` drops BOTH the logo and the divider (embedded chrome).
    ///   - animate: `false` freezes the dot pulse at peak (reduced-motion / snapshots).
    ///   - animateBars: opt-in ~80ms ease on each bar height (; default off).
    ///   - showsCancel: render the trailing ghost Cancel (default `false`, since the host
    ///     usually owns its own controls, law §9).
    ///   - cancelLabel: the ghost Cancel's label.
    ///   - accessibilityLabel: overrides the phase's combined live label.
    ///   - entrance: opt-in entrance/exit (; default `.none` = in place).
    ///   - onCancel: the ghost Cancel action.
    public init(
        phase: NockerlRecordingHudPhase = .recording,
        elapsedLabel: String = "",
        levels: [CGFloat] = [],
        errorMessage: String = "",
        transcribingLabel: String = "Transcribing…",
        pausedLabel: String = "Paused",
        resultPastedLabel: String = "Pasted",
        resultCopiedLabel: String = "Copied",
        showBrand: Bool = true,
        animate: Bool = true,
        animateBars: Bool = false,
        showsCancel: Bool = false,
        cancelLabel: String = "Cancel",
        accessibilityLabel: String? = nil,
        entrance: NockerlRecordingHudEntrance = .none,
        styleSelector: NockerlHudStyleSelector? = nil,
        onCancel: @escaping () -> Void = {}
    ) {
        self.phase = phase
        self.elapsedLabel = elapsedLabel
        self.levels = levels
        self.errorMessage = errorMessage
        self.transcribingLabel = transcribingLabel
        self.pausedLabel = pausedLabel
        self.resultPastedLabel = resultPastedLabel
        self.resultCopiedLabel = resultCopiedLabel
        self.showBrand = showBrand
        self.animate = animate
        self.animateBars = animateBars
        self.showsCancel = showsCancel
        self.cancelLabel = cancelLabel
        self.accessibilityLabelOverride = accessibilityLabel
        self.entrance = entrance
        self.styleSelector = styleSelector
        self.onCancel = onCancel
    }

    public var body: some View {
        if let styleSelector {
            // OPT-IN: the pill gains the expand-up drawer above it. The pill's OWN frame is
            // unchanged (§3.1). Only the panel above appears/collapses. The drawer view is
            // constructed ONLY in this branch (G1/G2: the base tree is never touched).
            VStack(spacing: NockerlSpace.space2) {
                if styleSelector.isOpen {
                    NockerlHudStyleDrawer(selector: styleSelector)
                        .transition(
                            reduceMotion
                                ? .opacity
                                : .scale(scale: 0.85, anchor: .bottom).combined(with: .opacity)
                        )
                }
                pill(restingStyle: styleSelector.showsResting ? styleSelector.activeStyle : nil)
            }
            .animation(
                reduceMotion ? nil : .nockerlStandard(NockerlMotionDuration.base),
                value: styleSelector.isOpen
            )
        } else {
            // BASE (default nil): byte-identical to today's HUD (no drawer, no VStack).
            pill(restingStyle: nil)
        }
    }

    /// The floating pill: the base HUD tree. `restingStyle` (non-nil only when a selector
    /// with `showsResting` is present) injects the disclosure chevron + active-style label
    /// into the constant-height content band; `nil` renders today's exact pill.
    @ViewBuilder
    private func pill(restingStyle: NockerlHudStyle?) -> some View {
        let palette = NockerlPalette.resolve(colorScheme)
        // The border rides the accent edge everywhere except error (statusWarning).
        let borderColor = phase == .error ? palette.statusWarning : palette.accentPrimary

        HStack(spacing: NockerlSpace.space3) {
            if showBrand {
                // The brand mark LEADS (task 2623), theme-adaptive ink, never cyan.
                NockerlLogo(size: 16)
                // The slight gray vertical divider (the web nk-hud__rule): 1×24.
                Rectangle()
                    .fill(palette.divider)
                    .frame(width: NockerlSpace.spacePx, height: NockerlSpace.space6)
            }

            // A CONSTANT content band (space6 = the divider height) so
            // EVERY phase renders at the same pill height; only WIDTH varies.
            content(palette: palette)
                .frame(height: NockerlSpace.space6)
                // v1.12.2 FIX 2: the recording readout (record-dot + count-up timer + wave)
                // WINS the horizontal space, so a long trailing style label WIDENS the pill
                // instead of squeezing the timer/wave (which overlapped + truncated the timer
                // to "…"). Inert in the base pill (it hugs its content, with no compression, so
                // base stays byte-identical, G2).
                .layoutPriority(1)

            if let restingStyle {
                // The resting style affordance: a trailing divider + the active-style
                // label. UNCONTROLLED (highlightedID == nil): a CLICKABLE `.plain` Button +
                // rotating chevron (mouse/keyboard open, v1.12.2). DRIVEN (v1.13.0,
                // highlightedID != nil): JUST the label, NO chevron, NO click affordance (the
                // non-activating pill is click-through; the host owns open/nav/commit, so a
                // chevron/keycap would falsely imply clickability).
                Rectangle()
                    .fill(palette.divider)
                    .frame(width: NockerlSpace.spacePx, height: NockerlSpace.space6)
                if styleSelector?.highlightedID == nil {
                    Button {
                        styleSelector?.isOpen.toggle()
                    } label: {
                        HStack(spacing: NockerlSpace.space1) {
                            Text(restingStyle.label)
                                .font(.nockerl(size: NockerlFontSize.size12, weight: .medium))
                                .foregroundColor(palette.onChromeMuted)
                                .lineLimit(1)
                            Image(systemName: "chevron.down")
                                .font(.system(size: NockerlFontSize.size10, weight: .semibold))
                                .foregroundColor(palette.onChromeMuted)
                                .rotationEffect(.degrees((styleSelector?.isOpen ?? false) ? 180 : 0))
                        }
                        .contentShape(Rectangle())
                    }
                    .buttonStyle(.plain)
                    .accessibilityLabel("Dictation style, \(restingStyle.label)")
                    .accessibilityHint("Opens the style menu")
                } else {
                    // DRIVEN: a passive label only (no chevron, no click; host-driven).
                    Text(restingStyle.label)
                        .font(.nockerl(size: NockerlFontSize.size12, weight: .medium))
                        .foregroundColor(palette.onChromeMuted)
                        .lineLimit(1)
                }
            }

            if showsCancel {
                NockerlButton(cancelLabel, variant: .ghost, size: .sm, action: onCancel)
            }
        }
        .padding(.vertical, NockerlSpace.space2)
        .padding(.horizontal, NockerlSpace.space4)
        .background(palette.chromeSurface)
        .clipShape(Capsule(style: .continuous))
        .overlay {
            // The §2 signature edge (accent), or statusWarning in the error phase.
            Capsule(style: .continuous)
                .strokeBorder(borderColor, lineWidth: NockerlFloatingBorder.width)
        }
        .overlay {
            // The lit-from-above top catch-light (the web's inset highlight).
            Capsule(style: .continuous)
                .strokeBorder(
                    LinearGradient(
                        colors: [palette.surfaceHighlight, .clear],
                        startPoint: .top,
                        endPoint: .center
                    ),
                    lineWidth: NockerlSpace.spacePx
                )
        }
        // L3 blur (raised/popover rung, matching the chat-input's float); the lift
        // opacity reads the shadowTintAlpha ladder at the matching level3 rung (0.33),
        // consistent with the L3 blur AND the Compose twin (which derives L3 from the rung).
        .shadow(
            color: palette.shadowTint.opacity(NockerlShadowTintAlpha.level3),
            radius: NockerlElevation.level3,
            x: 0,
            y: NockerlElevation.level3 / 2
        )
        .accessibilityElement(children: .contain)
        .accessibilityLabel(accessibilityLabelOverride ?? liveLabel)
        // Height is constant (the band above), so this animates only the
        // WIDTH: a smooth grow/shrink as the phase's content changes. Border-tone
        // and content also cross-fade on the same curve.
        .animation(.easeInOut(duration: 0.28), value: phase)
        // Opt-in entrance/exit: the host toggles presence in `withAnimation`; this
        // equips the pill to slide up from / down to the bottom edge (default none).
        .transition(
            entrance == .fromBottom
                ? .move(edge: .bottom).combined(with: .opacity)
                : .identity
        )
    }

    /// The morphing content after the logo + divider (one row per phase).
    @ViewBuilder
    private func content(palette: NockerlPalette) -> some View {
        switch phase {
        case .recording, .paused:
            let paused = phase == .paused
            HStack(spacing: NockerlSpace.space3) {
                // The ONE warm element: the record dot (statusError).
                Circle()
                    .fill(palette.statusError)
                    .frame(width: 8, height: 8)
                    .opacity(dotOpacity)
                    .onAppear {
                        guard animate, phase == .recording else { return }
                        withAnimation(.easeInOut(duration: 0.6).repeatForever(autoreverses: true)) {
                            pulsed = true
                        }
                    }

                // The timer: MONO + tabular (never jitters), THEME INK (never red).
                Text(elapsedLabel)
                    .font(.nockerlMono(size: NockerlFontSize.size12).monospacedDigit())
                    .foregroundColor(paused ? palette.onChromeMuted : palette.onCanvas)

                if !levels.isEmpty {
                    RecordingEqualizer(
                        levels: levels,
                        color: paused ? palette.onChromeMuted : palette.accentPrimary,
                        animateBars: animateBars && !paused
                    )
                }

                if paused {
                    statusWord(pausedLabel, palette: palette)
                }
            }

        case .transcribing:
            HStack(spacing: NockerlSpace.space2) {
                ProgressView()
                    .controlSize(.small)
                    .tint(palette.accentPrimary)
                statusWord(transcribingLabel, palette: palette)
            }

        case .error:
            HStack(spacing: NockerlSpace.space2) {
                // Color + ICON + TEXT (never color alone), the warm warning token.
                Image(systemName: "exclamationmark.triangle.fill")
                    .font(.system(size: 16))
                    .foregroundColor(palette.statusWarning)
                Text(errorMessage)
                    .font(.nockerl(size: NockerlFontSize.size12, weight: .regular))
                    .foregroundColor(palette.onChrome)
                    .lineLimit(1)
            }

        case let .result(pasted):
            HStack(spacing: NockerlSpace.space2) {
                // The completion glyph is BRAND cyan, not status-green: brand owns the
                // completion moment; the check glyph + Pasted/Copied label already carry the
                // success semantic, so color goes brand not status; unifies the HUD palette
                // (meters + §2 edge already cyan; green check was the lone outlier).
                Image(systemName: "checkmark.circle.fill")
                    .font(.system(size: 16))
                    .foregroundColor(palette.accentPrimary)
                statusWord(pasted ? resultPastedLabel : resultCopiedLabel, palette: palette)
            }
        }
    }

    /// The status word (Paused / Transcribing…): sans, medium, on-chrome ink.
    private func statusWord(_ text: String, palette: NockerlPalette) -> some View {
        Text(text)
            .font(.nockerl(size: NockerlFontSize.size12, weight: .medium))
            .foregroundColor(palette.onChrome)
            .lineLimit(1)
    }

    /// The record dot's opacity: paused static .5; recording pulses 1↔.3 (frozen 1).
    private var dotOpacity: Double {
        if phase == .paused { return 0.5 }
        if animate && pulsed { return 0.3 }
        return 1
    }

    /// The visually-hidden live label per phase (a11y: pill reads as status).
    private var liveLabel: String {
        switch phase {
        case .recording: return "Recording, \(elapsedLabel)"
        case .paused: return "Paused at \(elapsedLabel)"
        case .transcribing: return transcribingLabel
        case .error: return "Error: \(errorMessage)"
        case let .result(pasted): return pasted ? resultPastedLabel : resultCopiedLabel
        }
    }
}

/// The compact VU equalizer: 5 centered bars, log-normalized EXACTLY like both
/// shipped apps (`ln(raw+1)/ln(32768)`, raw = level × 32767, clamped to [0.08, 1]).
/// Geometry is LITERAL (barWidth 3, gap 3, height 20): identical across platforms,
/// not tokens. Only HEIGHT animates (opt-in ~80ms ease).
private struct RecordingEqualizer: View {
    let levels: [CGFloat]
    let color: Color
    let animateBars: Bool

    /// Log-normalize a raw 0…1 level to a bar height in points (the app math).
    private func barHeight(_ level: CGFloat) -> CGFloat {
        let raw = Double(max(0, min(level, 1))) * 32_767
        let normalized = log(raw + 1) / log(32_768)
        let clamped = max(0.08, min(1, normalized))
        return CGFloat(clamped) * 20
    }

    var body: some View {
        HStack(alignment: .center, spacing: 3) {
            ForEach(levels.indices, id: \.self) { index in
                Capsule(style: .continuous)
                    .fill(color)
                    .frame(width: 3, height: barHeight(levels[index]))
                    .animation(animateBars ? .easeOut(duration: 0.08) : nil, value: levels[index])
            }
        }
        .frame(height: 20)
    }
}

File: packages/compose/nockerl-design-components/src/main/kotlin/com/dizyx/nockerl/design/components/NockerlRecordingHud.kt
package com.dizyx.nockerl.design.components

import androidx.compose.animation.animateContentSize
import androidx.compose.animation.core.LinearOutSlowInEasing
import androidx.compose.animation.core.RepeatMode
import androidx.compose.animation.core.animateDpAsState
import androidx.compose.animation.core.animateFloat
import androidx.compose.animation.core.infiniteRepeatable
import androidx.compose.animation.core.rememberInfiniteTransition
import androidx.compose.animation.core.tween
import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.CheckCircle
import androidx.compose.material.icons.filled.Warning
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Icon
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.dizyx.nockerl.design.tokens.LocalNockerlColors
import com.dizyx.nockerl.design.tokens.NockerlElevation
import com.dizyx.nockerl.design.tokens.NockerlFloatingBorderWidth
import com.dizyx.nockerl.design.tokens.NockerlPillShape
import com.dizyx.nockerl.design.tokens.nockerlShadow
import kotlin.math.ln

/**
 * The **recording HUD**: the floating "Nockerl is listening" pill, aligned
 * VALUE-EXACT to the web canon (site RecordingHudDemo.tsx) per the design lead's
 * ruling: "look just like the web version." Twin of the Swift NockerlRecordingHUD.
 *
 * **Anatomy** (left → right, [showBrand] default true): `[ NockerlLogo (16) ] → [
 * gray vertical divider ] → [ CONTENT that morphs per [phase] ] → [ optional ghost
 * Cancel ]`.
 *
 * **The color split (the core fix: kills the over-warming, law §10):** the record
 * DOT is the ONLY warm element (statusError); the TIMER is THEME INK (onCanvas,
 * never red); the meter bars are accentPrimary (brand cyan). The pill carries the §2
 * floating accent edge (accentPrimary; statusWarning in the [phase] = error only).
 *
 * **Phases:** recording (pulsing dot → mono timer → live cyan meter) ·
 * paused (dimmed dot → muted timer → frozen muted meter → "Paused") · transcribing
 * (spinner → "Transcribing…") · error (warn triangle → text; border → statusWarning).
 *
 * **Phase motion:** HEIGHT IS CONSTANT across all phases (the content rides a
 * fixed 24.dp band); only WIDTH animates ([Modifier.animateContentSize]): a smooth
 * grow/shrink to each phase's content, never vertical movement. The OPT-IN
 * entrance/exit is host-owned (the compose idiom, law §9): wrap the HUD in
 * `AnimatedVisibility(visible, enter = slideInVertically { it } + fadeIn(), exit =
 * slideOutVertically { it } + fadeOut())`: it pops up from / shoots down to the
 * bottom edge, holding still while live. Default (no wrapper) = in place, unchanged.
 *
 * **Golden determinism:** pass [animate] = false to freeze the dot pulse at
 * peak; the meter draws [amplitudes] directly. The transcribing spinner is infinite
 * motion: document it, don't snapshot it (the search-field spinner precedent).
 *
 * @param modifier outer modifier (the host places this directly above the pill).
 * @param phase the dictation phase (drives the morphing content + border tone).
 * @param elapsedLabel the pre-formatted elapsed time (recording / paused).
 * @param amplitudes per-bar levels (0..1) for the equalizer (recording / paused).
 * @param errorMessage the failure text (the error phase).
 * @param transcribingLabel the transcribing status word.
 * @param pausedLabel the paused status word.
 * @param pasted RESULT phase: pasted-vs-copied, picks the success word.
 * @param resultPastedLabel the success word when [pasted] is true.
 * @param resultCopiedLabel the success word when [pasted] is false.
 * @param showBrand `false` drops BOTH the logo and the divider (embedded chrome).
 * @param animate `false` freezes the dot pulse at peak (deterministic goldens).
 * @param animateBars opt-in ~80ms ease on each bar height (default off).
 * @param showCancel render the trailing ghost Cancel (default false, the host
 *   usually owns its own controls, law §9).
 * @param cancelLabel the ghost cancel's label.
 * @param onCancel fired when the ghost Cancel is tapped.
 */
@Composable
fun NockerlRecordingHud(
    modifier: Modifier = Modifier,
    phase: NockerlRecordingHudPhase = NockerlRecordingHudPhase.RECORDING,
    elapsedLabel: String = "",
    amplitudes: List<Float> = emptyList(),
    errorMessage: String = "",
    transcribingLabel: String = "Transcribing…",
    pausedLabel: String = "Paused",
    pasted: Boolean = true,
    resultPastedLabel: String = "Pasted",
    resultCopiedLabel: String = "Copied",
    showBrand: Boolean = true,
    animate: Boolean = true,
    animateBars: Boolean = false,
    showCancel: Boolean = false,
    cancelLabel: String = "Cancel",
    onCancel: () -> Unit = {},
) {
    val colors = LocalNockerlColors.current
    // The border rides the accent edge everywhere except error (statusWarning).
    val borderColor = if (phase == NockerlRecordingHudPhase.ERROR) colors.statusWarning else colors.accentPrimary

    Surface(
        color = colors.chromeSurface,
        shape = NockerlPillShape,
        border = BorderStroke(NockerlFloatingBorderWidth, borderColor),
        // Height is constant (the content Box below), so this animates only
        // the WIDTH: a smooth grow/shrink as the phase content changes.
        modifier =
            modifier.animateContentSize().nockerlShadow(
                elevation = NockerlElevation.Level3,
                shape = NockerlPillShape,
            ),
    ) {
        Row(
            modifier = Modifier.padding(horizontal = 16.dp, vertical = 8.dp),
            horizontalArrangement = Arrangement.spacedBy(12.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            if (showBrand) {
                // The brand mark LEADS (task 2623), theme-adaptive ink, never cyan.
                NockerlLogo(size = 16.dp, contentDescription = "Nockerl")
                // The slight gray vertical divider (the web nk-hud__rule): 1×24.
                Box(
                    modifier =
                        Modifier
                            .width(1.dp)
                            .height(24.dp)
                            .background(colors.divider),
                )
            }
            // A CONSTANT content band (24.dp = the divider height) so EVERY
            // phase renders at the same pill height; only WIDTH varies.
            Box(modifier = Modifier.height(24.dp), contentAlignment = Alignment.CenterStart) {
                HudContent(
                    phase = phase,
                    elapsedLabel = elapsedLabel,
                    amplitudes = amplitudes,
                    errorMessage = errorMessage,
                    transcribingLabel = transcribingLabel,
                    pausedLabel = pausedLabel,
                    pasted = pasted,
                    resultPastedLabel = resultPastedLabel,
                    resultCopiedLabel = resultCopiedLabel,
                    animate = animate,
                    animateBars = animateBars,
                )
            }
            if (showCancel) {
                NockerlButton(
                    text = cancelLabel,
                    onClick = onCancel,
                    variant = NockerlButtonVariant.GHOST,
                )
            }
        }
    }
}

/** The morphing content after the logo + divider (one row per phase). */
@Composable
private fun HudContent(
    phase: NockerlRecordingHudPhase,
    elapsedLabel: String,
    amplitudes: List<Float>,
    errorMessage: String,
    transcribingLabel: String,
    pausedLabel: String,
    pasted: Boolean,
    resultPastedLabel: String,
    resultCopiedLabel: String,
    animate: Boolean,
    animateBars: Boolean,
) {
    val colors = LocalNockerlColors.current

    when (phase) {
        NockerlRecordingHudPhase.RECORDING, NockerlRecordingHudPhase.PAUSED -> {
            val paused = phase == NockerlRecordingHudPhase.PAUSED
            // Recording pulses the dot 1 → .3 on the web's 0.6s; paused holds .5.
            val dotAlpha =
                if (paused) {
                    0.5f
                } else if (animate) {
                    val transition = rememberInfiniteTransition(label = "nockerlRecPulse")
                    transition
                        .animateFloat(
                            initialValue = 1f,
                            targetValue = 0.3f,
                            animationSpec =
                                infiniteRepeatable(
                                    animation = tween(durationMillis = 600),
                                    repeatMode = RepeatMode.Reverse,
                                ),
                            label = "nockerlRecPulseAlpha",
                        ).value
                } else {
                    1f
                }

            Row(
                horizontalArrangement = Arrangement.spacedBy(12.dp),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                // The ONE warm element: the record dot (statusError).
                Box(
                    modifier =
                        Modifier
                            .size(8.dp)
                            .alpha(dotAlpha)
                            .clip(CircleShape)
                            .background(colors.statusError),
                )
                // The timer: MONO + tabular (never jitters), THEME INK (never red).
                Text(
                    text = elapsedLabel,
                    fontFamily = FontFamily.Monospace,
                    fontSize = 12.sp,
                    fontWeight = FontWeight.Medium,
                    color = if (paused) colors.onChromeMuted else colors.onCanvas,
                )
                if (amplitudes.isNotEmpty()) {
                    RecordingEqualizer(
                        amplitudes = amplitudes,
                        color = if (paused) colors.onChromeMuted else colors.accentPrimary,
                        animateBars = animateBars && !paused,
                    )
                }
                if (paused) {
                    StatusWord(pausedLabel)
                }
            }
        }

        NockerlRecordingHudPhase.TRANSCRIBING -> {
            Row(
                horizontalArrangement = Arrangement.spacedBy(8.dp),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                CircularProgressIndicator(
                    modifier = Modifier.size(14.dp),
                    color = colors.accentPrimary,
                    strokeWidth = 2.dp,
                )
                StatusWord(transcribingLabel)
            }
        }

        NockerlRecordingHudPhase.ERROR -> {
            Row(
                horizontalArrangement = Arrangement.spacedBy(8.dp),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                // Color + ICON + TEXT (never color alone), the warm warning token.
                Icon(
                    imageVector = Icons.Filled.Warning,
                    contentDescription = null,
                    tint = colors.statusWarning,
                    modifier = Modifier.size(16.dp),
                )
                Text(
                    text = errorMessage,
                    fontSize = 12.sp,
                    color = colors.onChrome,
                    maxLines = 1,
                )
            }
        }

        NockerlRecordingHudPhase.RESULT -> {
            Row(
                horizontalArrangement = Arrangement.spacedBy(8.dp),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                // The success glyph: statusSuccess (color + icon + text). No
                // dot/timer/meter, like transcribing.
                Icon(
                    imageVector = Icons.Filled.CheckCircle,
                    contentDescription = null,
                    tint = colors.statusSuccess,
                    modifier = Modifier.size(16.dp),
                )
                StatusWord(if (pasted) resultPastedLabel else resultCopiedLabel)
            }
        }
    }
}

/** The status word (Paused / Transcribing…), medium, on-chrome ink. */
@Composable
private fun StatusWord(text: String) {
    val colors = LocalNockerlColors.current
    Text(
        text = text,
        fontSize = 12.sp,
        fontWeight = FontWeight.Medium,
        color = colors.onChrome,
        maxLines = 1,
    )
}

/** The HUD's dictation phase: mirrors the web `HudPhase`. */
enum class NockerlRecordingHudPhase {
    /** Live capture: pulsing red dot → mono timer (ink) → live cyan meter. */
    RECORDING,

    /** Held: static dimmed dot → muted timer → frozen muted meter → "Paused". */
    PAUSED,

    /** Post-capture: indeterminate spinner → "Transcribing…". */
    TRANSCRIBING,

    /** Failed: warn triangle → error text; the pill border rides statusWarning. */
    ERROR,

    /** Success: a checkmark → "Pasted"/"Copied" (drops dot/timer/meter, like
     *  transcribing); the §2 accent edge stays. Pair with the `pasted` flag. */
    RESULT,
}

/**
 * The compact VU equalizer: 5 centered bars, log-normalized EXACTLY like both
 * shipped apps (`ln(raw+1)/ln(32768)`, raw = level × 32767, clamped to [0.08, 1]).
 * Geometry is LITERAL (barWidth 3, gap 3, height 20), identical across platforms.
 * Only HEIGHT animates (opt-in ~80ms ease).
 */
@Composable
private fun RecordingEqualizer(
    amplitudes: List<Float>,
    color: Color,
    animateBars: Boolean,
    modifier: Modifier = Modifier,
) {
    Row(
        modifier = modifier.height(20.dp),
        horizontalArrangement = Arrangement.spacedBy(3.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        amplitudes.forEach { amp ->
            val normalized = (ln(amp.coerceIn(0f, 1f) * 32_767.0 + 1.0) / ln(32_768.0)).toFloat()
            val target = (normalized.coerceIn(0.08f, 1f) * 20f).dp
            val barHeight =
                if (animateBars) {
                    animateDpAsState(
                        targetValue = target,
                        animationSpec = tween(durationMillis = 80, easing = LinearOutSlowInEasing),
                        label = "nockerlEqBar",
                    ).value
                } else {
                    target
                }
            Box(
                modifier =
                    Modifier
                        .width(3.dp)
                        .height(barHeight)
                        .clip(RoundedCornerShape(1.5.dp))
                        .background(color),
            )
        }
    }
}
